lib/protocol/protocol.go

```diff
@@ -421,6 +420,37 @@ func (c *rawConnection) readMessage() (message, error) {
 	return msg, nil
 }
 
+func (c *rawConnection) readHeader() (Header, error) {
+	// First comes a 2 byte header length
+
+	if len(c.readerBuf) < 2 {
```

Comment:
We never shrink the readerbuf (and other buffers), so it'll start out empty on the first message, grow to like two bytes then to a message size and then possibly at some point to a bigger message. We always use it like  `Read(c.readerBuf[:totalSize])` and similar, so no need to futz around with `cap()` etc.

That said, we *should* probably do a `getBuf(123)` as you say and have a sync.Pool in the background. As it is it's fairly efficient in that we're doing zero allocations once up and running, but on the other hand each connection will have four or so buffers permanently allocated, each sized for the biggest message ever seen.